### DISCUSSION TITLE
Infer component

### DISCUSSION THREAD

**Discussion by Aarebecca** [antvis/G2#4557] [Ideas]
## Component 可选组件

> ⚠️ Tooltip 不在此范围内
> Scrollable、Timeline 需要用户手动指定，不在推断范围中

<table>
  <thead>
    <tr>
      <th>组件</td>
      <th>细分</td>
      <th>说明</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan=3>Axis</td>
      <td>Linear</td>
      <td>直线坐标轴</td>
    </tr>
    <tr>
      <td>Arc</td>
      <td>弧线坐标轴</td>
    </tr>
    <tr>
      <td>Helix</td>
      <td>螺旋坐标轴(todo)</td>
    </tr>
    <tr>
      <td rowspan=2>Legend</td>
      <td>Category</td>
      <td>分类图例</td>
    </tr>
    <tr>
      <td>Continuous</td>
      <td>连续图例</td>
    </tr>
    <tr>
      <td rowspan=2>Scrollable</td>
      <td>Slider</td>
      <td>缩略轴</td>
    </tr>
    <tr>
      <td>Scrollbar</td>
      <td>滚动条</td>
    </tr>
    <tr>
      <td colspan=2>Timeline</td>
      <td>时间轴(todo)</td>
    </tr>
  </tbody>
</table>

## 组件推断

目前来说，组件推断主要依据三个条件：通道(`Channel`)、坐标系(`Coordinate`)、比例尺(`Scale`)。
例如坐标轴可以通过当前坐标系来确定是直角坐标系还是极坐标系下的轴类型，并依据通道、是否转置等条件进行进一步推断。

### Channel 通道

G2 5.0 目前可视相关的通道包括：

| 通道       | 说明   |
| ---------- | ------ |
| x, x1      |        |
| y, y1      |        |
| ~~series~~ |        |
| color      | 颜色   |
| size       | 尺寸   |
| opacity    | 透明度 |
| shape      | 形状   |

### 坐标系

G2 5.0 目前支持的坐标系和坐标系变换包括：

| 坐标系    | 说明                     |
| --------- | ------------------------ |
| cartesian | 直角坐标系(笛卡尔坐标系) |
| polar     | 极坐标系                 |
| parrallel | 平行坐标系               |
| helix     | 螺旋变换                 |
| theta     | 角度变换                 |
| radial    | 径向变换                 |
| transpose | 转置                     |
| reflect   | 镜像反转(沿中心点)       |
| reflectX  | 沿 X 轴反转              |
| reflectY  | 沿 Y 轴反转              |
| fisheye   | 鱼眼变换                 |
| fisheyeX  | 鱼眼变换(X 轴方向)       |
| fisheyeY  | 鱼眼变换(Y 轴方向)       |

### Scale 比例尺

目前 G2 5.0 提供了 14 种比例尺：

<table>
  <thead>
    <tr>
      <th>类型</th>
      <th>比例尺</th>
      <th>说明</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan=7>连续比例尺</td>
      <td>Linear</td>
      <td>线性比例尺</td>
    </tr>
    <tr>
      <td>Identity</td>
      <td>等比比例尺</td>
    </tr>
    <tr>
      <td>Log</td>
      <td>对数比例尺</td>
    </tr>
    <tr>
      <td>Pow</td>
      <td>指数比例尺</td>
    </tr>
    <tr>
      <td>Sqrt</td>
      <td>平方数比例尺</td>
    </tr>
    <tr>
      <td>Sequential</td>
      <td>连续比例尺</td>
    </tr>
    <tr>
      <td>Time</td>
      <td>时间比例尺</td>
    <tr>
      <td rowspan=3>分类比例尺</td>
      <td>Threshold</td>
      <td>阈值比例尺</td>
    </tr>
    <tr>
      <td>Quantize</td>
      <td>等比例比例尺</td>
    </tr>
    <tr>
      <td>Quantile</td>
      <td>等分位比例尺</td>
    </tr>
    <tr>
      <td rowspan=3>离散比例尺</td>
      <td>Ordinal</td>
      <td>序数比例尺</td>
    </tr>
    <tr>
      <td>Band</td>
      <td>分段比例尺</td>
    </tr>
    <tr>
      <td>Point</td>
      <td>点比例尺</td>
    </tr>
  </tbody>
</table>

### 组件类型推断

**基于`通道`+`坐标系`推断**

可推断组件：`Axis`

| 坐标系\\通道 | x       | x1      | y       | y1      | 说明     |
| ------------ | ------- | ------- | ------- | ------- | -------- |
| 直角坐标系   | `axisX` | `axisX` | `axisY` | `axisY` | 转置取反 |
| 极坐标系     | `arc`   | `arc`   | `axisY` | `axisY` |          |
| 平行坐标系   | `axisX` |         | `axisY` |         | 转置取反 |

**基于`通道`+`比例尺`推断**

可推断组件：`category`、`continuous`，其中连续图例根据类型可细分为：

- `continuous` 默认类型
- `continuousSize` 尺寸类型
- `continuousBlock` 分块类型
- `continuousSizeBlock` 尺寸分块类型

| 比例尺\\通道 | color             | size             | opacity           | shape      |
| ------------ | ----------------- | ---------------- | ----------------- | ---------- |
| 连续比例尺<br/>(不含时间比例尺)   | `continuous`      | `continuousSize` | `continuous`      | /          |
| 分类比例尺   | `continuousBlock` | `continuousBlockSize` | `continuousBlock` | /          |
| 离散比例尺   | `category`        | `category`       | `category`        | `category` |

## 组件位置

`top`、`bottom`、`left`、`right` 即可描述大多数组件相对于可视化主体元素的位置。
> ⚠️ 直角坐标系下的坐标轴理论上应当于其所描述的元素对齐。
> 在极坐标系下，需要考虑`Axis`是位于`环内`还是`环外`

```ts
type PrimitivePosition = 'top' | 'bottom' | 'left' | 'right' | 'center' | 'inner' | 'outer';
type Position = PrimitivePosition | `${PrimitivePosition}-${PrimitivePosition}`
```

<img width="200" alt="image" src="https://user-images.githubusercontent.com/25787943/211194683-8382d1b0-41a5-4ad5-8e5b-9b0d3475eaef.png">

此外，大多数组件是具有方向性的，即水平朝向`Horizontal`和垂直朝向`Vertical`。

> ⚠️ 对于一些组件来说，可选的方向与其所在的位置是有密切关联的
> 例如直角坐标系下位于`left`、`right`位置的坐标轴、缩略轴和滚动条，其方向一定是`vertical`朝向的，因此对这类组件设置`Orientation`则不会生效
> 此外， `Orientation` 支持设定一个数值来指定其朝向。

```ts
type Orientation = 'Horizontal' | 'Vertical' | number;
```

> ⚠️ 平行坐标系下，`Anchor`、`Align`、`Orientation`对坐标轴组件无效。

> ⚠️ 极坐标系下通常同时具有直线坐标轴和弧线坐标轴，配置`Align`对直线坐标轴无效，
> 配置`Orientation`对弧线坐标轴无效。

### 组件位置推断

**`Axis` 位置推断**

> 坐标轴的 `Position` 通常来说仅支持基本的 Primitive Position , 不支持复合 Position
> ⚠️ 极坐标系下 linear axis 还取决于其角度，若在 `angle % 180 === 0` 则应当为 `axisX`, `angle % 90 === 0` 则为 `axisY`。

<table>
  <thead>
    <tr>
      <td>坐标系</td>
      <td>Position</td>
      <td>Orientation</td>
      <td>组件/(转置时)</td>
      <td>说明</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan=2>直角坐标系</td>
      <td>left / right</td>
      <td>-</td>
      <td>axisY (axisX)</td>
      <td>直线坐标轴</td>
    </tr>
    <tr>
      <td>top / bottom</td>
      <td>-</td>
      <td>axisX (axisY)</td>
      <td>直线坐标轴</td>
    </tr>
    <tr>
      <td rowspan=2>极坐标系</td>
      <td>center</td>
      <td>'horizontal' | 'vertical' | number</td>
      <td>axisY (axisX)</td>
      <td>直线坐标轴</td>
    </tr>
    <tr>
      <td>inner / outer</td>
      <td>-</td>
      <td>axisArc</td>
      <td>弧线坐标轴</td>
    </tr>
     <tr>
      <td rowspan=2>平行坐标系</td>
      <td>center</td>
      <td>'horizontal' | 'vertical' </td>
      <td>axisX / axisY</td>
      <td>常规平行坐标图</td>
    </tr>
     <tr>
      <td>center</td>
      <td>number</td>
      <td>axisLinear</td>
      <td>雷达图</td>
    </tr>
  </tbody>
</table>

**通用组件位置推断**

组件位置默认为：

```ts
{
  position: 'top',
  orientation: 'horizontal'
}
```

其余位置需要用户手动指定。

**Comment by pearmini** [comment]
这个梳理的挺好的，等 2 月份重构的时候可以再深入讨论一下
